Add CLI config:get and config:delete

mwithheld opened the pull request:
### Description:

Issue #16576 Add logic to CLI for config:delete and config:get

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

mwithheld committed:
16576 Add ConfigGet and tests

mwithheld committed:
15676 config:get code and test updates

mwithheld committed:
15676 Add config:delete code and tests

mwithheld committed:
15676 Add config:delete code and tests

mwithheld committed:
Merge branch 'matomo-org:4.x-dev' into 16576

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigDelete.php:
```diff
@@ -0,0 +1,319 @@
+#Delete this setting.
+$ ./console %command.name% --section=database --key=username
+#Delete one value in an array:
+$ ./console %command.name% --section=database --key=username --value=DevicesDetection
```
for consistency:
```suggestion
$ ./console %command.name% --section=PluginsInstalled --key=PluginsInstalled --value=DevicesDetection
```

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigDelete.php:
```diff
@@ -0,0 +1,319 @@
+                break;
+            default:
+                // We should not get here, but just in case.
+                throw new Exception('Some unexpected error occurred parsing input values');
```
```suggestion
                throw new \Exception('Some unexpected error occurred parsing input values');
```

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigDelete.php:
```diff
@@ -0,0 +1,319 @@
+            'key' => $input->getOption('key'),
+            'value' => $input->getOption('value'),
+        ]);
+        $debug && fwrite(STDERR, PHP_EOL . __FUNCTION__ . '::Started with $options=' . (empty($options) ? '' : serialize($options)));
```
Personally I would vote for removing all those debug code. It actually makes the code bit harder to read. And imho debugging the code has more benefits than using plenty of debug output.
Not sure if maybe @tsteur has another opinion on that.

mwithheld replied:
I have removed the debug code.

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigDelete.php:
```diff
@@ -0,0 +1,319 @@
+        // Check the setting exists and user has permissions. If so, put it in the wrapper.
+        switch (true) {
+            case ! $settingWrapped->setting->isWritableByCurrentUser():
+                throw new \Exception('No write permissions to this setting');
```
Console commands are always executed as super user, so there is actually no need to check for permission.

mwithheld replied:
Good point - done.

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigDelete.php:
```diff
@@ -0,0 +1,319 @@
+            }
+        }
+
+        //Many matomo script output Done when they're done.  IMO it's not needed: $output->writeln(self::wrapInTag('info', 'Done.'));
```
The command prints a success or error message already. So no need to print anything else.
```suggestion
```

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigDelete.php:
```diff
@@ -0,0 +1,319 @@
+        // Sanity check inputs.
+        switch (true) {
+            case empty($argument) && empty($options):
+                throw new \InvalidArgumentException('You must set either an argument or set options --section and optional --key');
```
Guess it should be this:
```suggestion
                throw new \InvalidArgumentException('You must set either an argument or set options --section, --key and optional --value');
```

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigDelete.php:
```diff
@@ -0,0 +1,319 @@
+        if ($settingWrappedNew->isArray) {
+            $debug && fwrite(STDERR, PHP_EOL . __FUNCTION__ . "This config setting is an array");
+            if (empty($settingWrappedNew->arrayVal)) {
+                throw new \InvalidArgumentException('This config setting an array, and no array value was specified for deletion');
```
```suggestion
                throw new \InvalidArgumentException('This config setting is an array, but no array value was specified for deletion');
```

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigDelete.php:
```diff
@@ -0,0 +1,319 @@
+        $this->addArgument(
+            'argument',
+            InputArgument::OPTIONAL,
+            "A config setting in the format Section.key or Section.array_key[], e.g. 'Database.username' or 'PluginsInstalled.PluginsInstalled[CustomDimensions]"
```
```suggestion
            "A config setting in the format Section.key or Section.array_key[], e.g. 'Database.username' or 'PluginsInstalled.PluginsInstalled.CustomDimensions'"
```

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigGet.php:
```diff
@@ -0,0 +1,343 @@
+    {
+        $result = json_encode($var, JSON_UNESCAPED_SLASHES);
+        if ($result === false) {
+            throw new Exception('Failed to json_encode');
```
```suggestion
            throw new \Exception('Failed to json_encode');
```

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigGet.php:
```diff
@@ -0,0 +1,343 @@
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Yaml\Yaml;
```
This might actually only be a dev dependency, so it might not be available in releases. But you might use `\Spyc` instead, which is included through the device detector dependency.

mwithheld replied:
Thanks - Symfony Yaml is not in the pre-packaged download, anyway.  Changed to use Spyc.

sgiehl commented on plugins/CoreAdminHome/Commands/ConfigGet.php:
```diff
@@ -0,0 +1,343 @@
+        //Optionally could set this at runtime with: $output->getVerbosity() >= OutputInterface::VERBOSITY_VERBOSE.
+        $debug = false;
+
+        // Gather options, then discard ones with an empty value so we do not need to check for empty later.
+        $options = array_filter([
+            'section' => $input->getOption('section'),
+            'key' => $input->getOption('key'),
+            'format' => $input->getOption('format'),
+        ]);
+        $debug && fwrite(STDERR, PHP_EOL . __FUNCTION__ . '::Started with $options=' . (empty($options) ? '' : serialize($options)));
+
+        // If none specified, set default format.
+        if ( ! isset($options['format']) || empty($format = $options['format']) || ! in_array($format, self::OUTPUT_FORMATS, true)) {
+            $format = self::OUTPUT_FORMAT_DEFAULT;
+        }
+        // Since it is config and not input data, remove format from the options so we can just work with the section and format.
+        unset($options['format']);
```
Maybe handling the format independently makes the code a bit cleaner
```suggestion
        // Gather options, then discard ones with an empty value so we do not need to check for empty later.
        $options = array_filter([
            'section' => $input->getOption('section'),
            'key' => $input->getOption('key'),
        ]);

        // If none specified, set default format.
        $format = $input->getOption('format');
        if ( empty($format) || ! in_array($format, self::OUTPUT_FORMATS, true)) {
            $format = self::OUTPUT_FORMAT_DEFAULT;
        }
```

sgiehl commented on plugins/CoreAdminHome/tests/Integration/ConfigDeleteTest.php:
```diff
@@ -0,0 +1,425 @@
+        self::removeTestConfigFile();
+        parent::setUp();
+    }
+
+    public function tearDown(): void
+    {
+        parent::tearDown();
+    }
```
could be removed, as not needed.
```suggestion
```

sgiehl commented on plugins/CoreAdminHome/tests/Integration/ConfigDeleteTest.php:
```diff
@@ -0,0 +1,425 @@
+        $inputArr = [
+            'command' => self::COMMAND,
+            '-vvv' => false,
+            'argument' => $sectionName . $settingName . (empty($value) ? '' : ".$value"),
```
This would actually need to be this in order to work:
```suggestion
            'argument' => $sectionName . '.' . $settingName . (empty($value) ? '' : ".$value"),
```

mwithheld replied:
Good catch.  Fixed.

sgiehl commented on plugins/CoreAdminHome/tests/Integration/ConfigDeleteTest.php:
```diff
@@ -0,0 +1,425 @@
+        $debug = false;
+
+        // Pass empty section name.
+        $resultObj = $this->runCommandWithArguments('');
```
Was it on purpose to only have one test running with arguments, that actually only passes an empty one? Maybe might be useful to also have some working tests and maybe some edge cases. (Same applies for the ConfigGetTest)

mwithheld replied:
They just fell off my radar.  Added tests with args to ConfigGetTest and ConfigDeleteTest classes.

sgiehl reviewed (changes requested):
@mwithheld Thanks for creating this pull request. I did a review and left a couple of comments. Would be nice if you have some time to apply that

mwithheld committed:
Improve usage examples

Co-authored-by: Stefan Giehl <stefan@matomo.org>

mwithheld committed:
Fix Exception namespace

Co-authored-by: Stefan Giehl <stefan@matomo.org>

mwithheld committed:
Remove comment about Done output

Co-authored-by: Stefan Giehl <stefan@matomo.org>

mwithheld committed:
Clarify parsing format option

Co-authored-by: Stefan Giehl <stefan@matomo.org>

mwithheld committed:
Remove unneeded teardown()

Co-authored-by: Stefan Giehl <stefan@matomo.org>

mwithheld committed:
Remove isWritableByCurrentUser() check - CLI is always superuser

mwithheld committed:
Use Spyc yaml included in releases not Symfony

mwithheld committed:
PSR12 Fixes

# Conflicts:
#	plugins/CoreAdminHome/Commands/ConfigDelete.php
#	plugins/CoreAdminHome/Commands/ConfigGet.php

mwithheld committed:
PSR12 Fixes

mwithheld committed:
runCommandWithArguments: Fix missing dot; remove adding value param

mwithheld committed:
Fix paths in single-test usage example

mwithheld committed:
Add tests using args

mwithheld committed:
If there are multiple arguments just use the last one

mwithheld committed:
Add tests using args

mwithheld committed:
Remove debug code

mwithheld committed:
Restore accidentally deleted testUsingOptsNonExistentSectionAndSettingShouldYieldEmpty

mwithheld committed:
Remove debug comment and blank lines

mwithheld committed:
Fix testUsingArgs should use runCommandWithArguments()

sgiehl committed:
Apply remaining suggestions from code review

sgiehl committed:
Update and rename plugins/CoreAdminHome/tests/Integration/ConfigDeleteTest.php to plugins/CoreAdminHome/tests/Integration/Commands/ConfigDeleteTest.php

sgiehl committed:
Update and rename plugins/CoreAdminHome/tests/Integration/ConfigGetTest.php to plugins/CoreAdminHome/tests/Integration/Commands/ConfigGetTest.php

sgiehl committed:
Update and rename plugins/CoreAdminHome/tests/Integration/SetConfigTest.php to plugins/CoreAdminHome/tests/Integration/Commands/SetConfigTest.php

sgiehl reviewed (approved):
Thanks again @mwithheld for applying the changes. I've added some last fixes and now it should be good to merge 🎉
